services/recovery_method.js

```diff
@@ -1,69 +1,215 @@
-        return SequelizeRecoveryMethods.createRecoveryMethod({
-            accountId,
-            detail,
-            kind,
-            publicKey,
-            requestId,
-            securityCode,
-        });
+const RecoveryMethodService = stampit({
+    props: {
+        db: {
+            createRecoveryMethod,
+            deleteRecoveryMethod,
+            getRecoveryMethodByIdentity,
+            listRecoveryMethodsByAccountId,
+            updateRecoveryMethod,
+        },
+        sequelize: SequelizeRecoveryMethods,
     },
+    methods: {
+        createRecoveryMethod({ accountId, detail, kind, publicKey, requestId, securityCode }) {
+            if (!USE_DYNAMODB) {
+                return this.sequelize.createRecoveryMethod({
+                    accountId,
+                    detail,
+                    kind,
+                    publicKey,
+                    requestId,
+                    securityCode,
+                });
+            }
+            return this.db.createRecoveryMethod({
+                accountId,
+                detail,
+                kind,
+                publicKey,
+                requestId,
+                securityCode,
+            });
+        },
 
-    deleteOtherRecoveryMethods({ accountId, detail }) {
-        return SequelizeRecoveryMethods.deleteOtherRecoveryMethods({ accountId, detail });
-    },
+        deleteOtherRecoveryMethods({ accountId, detail }) {
+            if (!USE_DYNAMODB) {
+                return this.sequelize.deleteOtherRecoveryMethods({ accountId, detail });
+            }
+            return this.db.listRecoveryMethodsByAccountId(accountId)
+                .filter((recoveryMethod) => recoveryMethod.detail !== detail)
+                .map((recoveryMethod) => deleteRecoveryMethod(recoveryMethod));
+        },
 
-    deleteRecoveryMethod({ accountId, kind, publicKey }) {
-        return SequelizeRecoveryMethods.deleteRecoveryMethod({ accountId, kind, publicKey });
-    },
+        deleteRecoveryMethod({ accountId, kind, publicKey }) {
+            if (!USE_DYNAMODB) {
+                return this.sequelize.deleteRecoveryMethod({ accountId, kind, publicKey });
+            }
+            return this.db.listRecoveryMethodsByAccountId(accountId)
+                .filter((recoveryMethod) => recoveryMethod.kind === kind && recoveryMethod.publicKey === publicKey)
+                .map((recoveryMethod) => deleteRecoveryMethod(recoveryMethod));
+        },
 
-    getTwoFactorRecoveryMethod(accountId) {
-        return SequelizeRecoveryMethods.getTwoFactorRecoveryMethod(accountId);
-    },
+        getTwoFactorRecoveryMethod(accountId) {
+            if (!USE_DYNAMODB) {
+                return this.sequelize.getTwoFactorRecoveryMethod(accountId);
+            }
+            return this.db.listRecoveryMethodsByAccountId(accountId)
+                .filter((recoveryMethod) => recoveryMethod.kind.startsWith('2fa-'))
+                .get(0)
```

Comment:
Is it only possible to have one 2fa recovery method?

Reply:
In theory, yes. You can only have one of any type of recovery method (and in the case of 2FA only one of `2fa-phone` or `2fa-email`). In practice there are duplicates 😞